File: typegraph/python/typegraph/runtimes/kv.py
# Copyright Metatype OÜ, licensed under the Mozilla Public License Version 2.0.
# SPDX-License-Identifier: MPL-2.0

from dataclasses import dataclass
from typing import Optional

from typegraph import fx, t
from typegraph.gen.exports.runtimes import (
    BaseMaterializer,
    Effect,
    KvMaterializer,
    KvRuntimeData,
)
from typegraph.gen.types import Err
from typegraph.runtimes.base import Materializer, Runtime
from typegraph.wit import runtimes, store


class KvRuntime(Runtime):
    host: str
    port: Optional[str]
    db_number: Optional[int]
    password: Optional[str]

    def __init__(
        self,
        host: str,
        port: Optional[str],
        db_number: Optional[int],
        password: Optional[str],
    ):
        data = KvRuntimeData(
            host=host, port=port, db_number=db_number, password=password
        )
        runtime_id = runtimes.register_kv_runtime(store, data)
        if isinstance(runtime_id, Err):
            raise Exception(runtime_id.value)

        super().__init__(runtime_id.value)
        self.host = host
        self.port = port
        self.db_number = db_number
        self.password = password

    def get(self):
        mat = self.__operation(KvMaterializer.GET, fx.read())
        return t.func(t.struct({"key": t.string()}), t.string(), mat)

    def set(self):
        mat = self.__operation(KvMaterializer.SET, fx.update())
        return t.func(
            t.struct({"key": t.string(), "value": t.string()}), t.string(), mat
        )

    def delete(self):
        mat = self.__operation(KvMaterializer.DELETE, fx.delete())
        return t.func(t.struct({"key": t.string()}), t.string(), mat)

    def keys(self):
        mat = self.__operation(KvMaterializer.KEYS, fx.read())
        return t.func(
            t.struct({"filter": t.optional(t.string())}), t.list(t.string()), mat
        )

    def all(self):
        mat = self.__operation(KvMaterializer.ALL, fx.read())
        return t.func(
            t.struct({"filter": t.optional(t.string())}),
            t.list(t.string()),
            mat,
        )

    def __operation(self, operation: KvMaterializer, effect: Effect):
        mat_id = runtimes.kv_operation(
            store, BaseMaterializer(self.id, effect), operation
        )
        if isinstance(mat_id, Err):
            raise Exception(mat_id.value)

        return KvOperationMat(mat_id.value, effect=effect, operation=operation)


@dataclass
class KvOperationMat(Materializer):
    operation: KvMaterializer

File: typegraph/deno/sdk/src/runtimes/kv.ts
// Copyright Metatype OÜ, licensed under the Mozilla Public License Version 2.0.
// SPDX-License-Identifier: MPL-2.0

import { Materializer, Runtime } from "./mod.ts";
import * as t from "../types.ts";
import { runtimes } from "../wit.ts";
import { Effect, KvMaterializer } from "../gen/typegraph_core.d.ts";

import { fx } from "../index.ts";

class KvOperationMat implements Materializer {
  operation: KvMaterializer;
  _id: number;
  constructor(id: number, operation: KvMaterializer) {
    this._id = id;
    this.operation = operation;
  }
}

export class KvRuntime extends Runtime {
  host: string;
  port?: string;
  dbNumber?: number;
  password?: string;

  constructor(
    { host, port, dbNumber, password }: {
      host: string;
      port?: string;
      dbNumber?: number;
      password?: string;
    },
  ) {
    const id = runtimes.registerKvRuntime({ host, port, dbNumber, password });
    super(id);
    this.host = host;
    this.port = port;
    this.dbNumber = dbNumber;
    this.password = password;
  }

  #operation(operation: KvMaterializer, effect: Effect) {
    const mad_id = runtimes.kvOperation({
      runtime: this._id,
      effect: effect,
    }, operation);

    return new KvOperationMat(mad_id, operation);
  }

  set() {
    const mat = this.#operation("set", fx.update());
    return t.func(
      t.struct({ "key": t.string(), "value": t.string() }),
      t.string(),
      mat,
    );
  }

  get() {
    const mat = this.#operation("get", fx.read());

    return t.func(t.struct({ "key": t.string() }), t.string(), mat);
  }

  delete() {
    const mat = this.#operation("delete", fx.delete_());

    return t.func(t.struct({ "key": t.string() }), t.string(), mat);
  }

  keys() {
    const mat = this.#operation("keys", fx.read());

    return t.func(
      t.struct({ "filter": t.optional(t.string()) }),
      t.list(t.string()),
      mat,
    );
  }
  all() {
    const mat = this.#operation("all", fx.read());

    return t.func(
      t.struct({ "filter": t.optional(t.string()) }),
      t.list(t.string()),
      mat,
    );
  }
}
